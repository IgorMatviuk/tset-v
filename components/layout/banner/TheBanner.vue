<template>
  <div
    :class="{
      'top-info--animate': isHomePage,
      'top-info--animated': isAnimationFinished
    }"
    class="top-info"
  >
    <BContainer class="top-info__container">
      <NuxtLink
        :to="$i18n.path(banner.url)"
        class="top-info__link"
        target="_blank"
      />
      <picture>
        <source
          srcset="/img/stub.png"
          :data-srcset="banner.image.large_image_url"
          type="image/webp"
          media="(min-width: 1200px)"
        />
        <source
          srcset="/img/stub.png"
          :data-srcset="banner.image.large_image_url"
          type="image/png"
          media="(min-width: 1200px)"
        />
        <source
          srcset="/img/stub.png"
          :data-srcset="banner.image.medium_image_url"
          type="image/webp"
          media="(min-width: 768px)"
        />
        <source
          srcset="/img/stub.png"
          :data-srcset="banner.image.medium_image_url"
          type="image/png"
          media="(min-width: 768px)"
        />
        <source
          srcset="/img/stub.png"
          :data-srcset="banner.image.small_image_url"
          type="image/webp"
        />
        <img
          :data-src="banner.image.small_image_url"
          src="/img/stub.png"
          data-object-fit="cover"
          data-object-position="center bottom"
          alt="Top Info"
          class="top-info__img lazyload"
          width="100"
          height="100"
          loading="lazy"
        />
      </picture>
    </BContainer>
  </div>
</template>

<script>
import gsap from 'gsap'
import { mapState } from 'vuex'

export default {
  props: {
    banner: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      isAnimationFinished: false
    }
  },
  computed: {
    ...mapState({
      isBackgroundRevealed: (state) => state.ui.isBackgroundRevealed,
      currLocale: (state) => state.locale
    }),
    isHomePage() {
      return (
        this.$route.path === '/' ||
        this.$route.path === `/${this.currLocale}` ||
        this.$route.path === `/${this.currLocale}/`
      )
    }
  },
  watch: {
    isBackgroundRevealed: 'animate'
  },
  mounted() {
    if (!this.isHomePage) {
      this.isAnimationFinished = true
    }
  },
  methods: {
    animate(status) {
      /* istanbul ignore else */
      if (status && this.isHomePage) {
        gsap.to(this.$el, {
          y: 0,
          duration: 1,
          delay: 0.15,
          ease: 'expo.out(1, 0.3)',
          onComplete: () => (this.isAnimationFinished = true)
        })
      }
    }
  }
}
</script>

<style lang="scss">
.top-info {
  position: relative;
  z-index: 12;
  height: 30px;
  background-color: $primary-light-1;

  &__container {
    position: relative;
  }

  &__link {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }

  &__img {
    width: 100%;
    height: 30px;
    object-fit: cover;
    object-position: center;
  }
}

.has-banner .top-info--animate:not(.top-info--animated) {
  @include media-breakpoint-up(lg) {
    transform: translateY(-30px);
  }
}
</style>
