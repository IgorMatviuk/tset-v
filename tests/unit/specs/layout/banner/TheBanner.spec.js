import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import { BContainer } from 'bootstrap-vue'

import TheBanner from '@/components/layout/banner/TheBanner'

import {
  mutationTypes as uiMutationTypes,
  mutations as uiMutations
} from '@/store/ui'

import banner from '@/api/fixtures/banner'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.component('BContainer', BContainer)

describe('TheBanner', () => {
  let store

  function createBanner({ path = '/' } = {}) {
    return shallowMount(TheBanner, {
      localVue,
      store,
      propsData: { banner },
      mocks: {
        $route: { path },
        $i18n: {
          path: (path) => path
        }
      },
      stubs: ['NuxtLink']
    })
  }

  beforeEach(() => {
    store = new Vuex.Store({
      state: {
        locale: 'ua'
      },
      modules: {
        ui: {
          namespaced: true,
          state: {
            isBackgroundRevealed: false
          },
          mutations: { ...uiMutations }
        }
      }
    })
  })

  test('has expected structure', () => {
    const wrapper = createBanner()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('triggers animation on backgroud reveal', () => {
    const wrapper = createBanner()

    expect(wrapper.vm.isBackgroundRevealed).toBe(false)
    store.commit(`ui/${uiMutationTypes.SET_BACKGROUND_REVEALED_TOGGLE}`, true)
    expect(wrapper.vm.isBackgroundRevealed).toBe(true)
  })

  describe('Homepage detection', () => {
    test('default path', () => {
      const wrapper = createBanner({ path: `/ua` })
      expect(wrapper.vm.isHomePage).toBe(true)
    })

    test('path with trailing slash', () => {
      const wrapper = createBanner({ path: `/ua/` })
      expect(wrapper.vm.isHomePage).toBe(true)
    })
  })
})
