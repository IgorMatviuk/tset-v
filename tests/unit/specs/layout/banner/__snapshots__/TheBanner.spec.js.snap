// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheBanner has expected structure 1`] = `
<div
  class="top-info top-info--animate"
>
  <bcontainer-stub
    class="top-info__container"
    tag="div"
  >
    <nuxtlink-stub
      class="top-info__link"
      target="_blank"
      to="bryuki-i-shorty"
    />
     
    <picture>
      <source
        data-srcset="/img/top-info/top-info_xl.png"
        media="(min-width: 1200px)"
        srcset="/img/stub.png"
        type="image/webp"
      />
       
      <source
        data-srcset="/img/top-info/top-info_xl.png"
        media="(min-width: 1200px)"
        srcset="/img/stub.png"
        type="image/png"
      />
       
      <source
        data-srcset="/img/top-info/top-info_md.png"
        media="(min-width: 768px)"
        srcset="/img/stub.png"
        type="image/webp"
      />
       
      <source
        data-srcset="/img/top-info/top-info_md.png"
        media="(min-width: 768px)"
        srcset="/img/stub.png"
        type="image/png"
      />
       
      <source
        data-srcset="/img/top-info/top-info_sm.png"
        srcset="/img/stub.png"
        type="image/webp"
      />
       
      <img
        alt="Top Info"
        class="top-info__img lazyload"
        data-object-fit="cover"
        data-object-position="center bottom"
        data-src="/img/top-info/top-info_sm.png"
        height="100"
        loading="lazy"
        src="/img/stub.png"
        width="100"
      />
    </picture>
  </bcontainer-stub>
</div>
`;
